Added interrupt handler ordering tests.

jpd002 opened the pull request:
Added some tests to see how the EE kernel handles interrupt (INTC and DMAC) handler ordering.

jpd002 committed:
Added interrupt handler ordering tests.

unknownbrackets commented on tests/kernel/ee/interrupt/dmac_order.cpp:
```diff
@@ -0,0 +1,169 @@
+	g_dmacHandlerCount++;
+}
+
+void removeTestDmacHandler(int index) {
```
Hmm.  Should `addTestDmacHandler` return an id that is used here?  I feel like the `removeTestDmacHandler(1)` along with calling the functions 1/2/3 with a template is confusing, since they're differently indexed.

-[Unknown]

jpd002 replied:
I agree it's a bit confusing, but I wanted `removeTestDmacHandler` to remove the n-th handler in the array. Doing the removal by id would require an additional step to go through the array and find where it's located.

But if you think it's better, I don't mind changing it so `addTestDmacHandler` returns an id and `removeTestDmacHandler` removes the handler with specified id from the array.

I also wish I could use a map or something here, because implementing the array management functions is no fun. The STL doesn't seem to work very well in PS2SDK, so maybe it would be interesting to build some simple containers at some point?

unknownbrackets replied:
Fair enough.  Yeah, I don't think I tried STL yet for fear that they likely would not work.  I may have time later this month to investigate that... it would definitely be useful.

-[Unknown]

unknownbrackets commented on tests/kernel/ee/interrupt/dmac_order.cpp:
```diff
@@ -0,0 +1,169 @@
+		*(R_EE_D9_SADR) = 0;
+		*(R_EE_D9_QWC)  = 1;
+		*(R_EE_D9_CHCR) = 0x100;
+		while(*(R_EE_D9_CHCR) & 0x100) {}
```
Please use `continue;` for a `while` without a body, with regular formatting, e.g.:

```
while (*(R_EE_D9_CHCR) & 0x100) {
    continue;
}
```

-[Unknown]

jpd002 replied:
Sounds good, I will make that change.

unknownbrackets commented on tests/kernel/ee/interrupt/dmac_order.cpp:
```diff
@@ -0,0 +1,169 @@
+	pushFrontTest();
+	mixedFrontBackTest();
+	removeTest();
+	relativeTest();
```
It's pedantic, sorry, but I think you should swap the order of remove/relative either here or above so they match.

-[Unknown]

jpd002 replied:
Agreed, I will change it.

unknownbrackets commented on tests/kernel/ee/interrupt/dmac_order.cpp:
```diff
@@ -0,0 +1,169 @@
+template <s32 number>
+s32 cancelDmacHandler(s32 channel) {
+	appendResult(number);
+	return -1;
```
Is this behavior the same for all negative values (e.g. 0x80000000)?  What about non-zero (like 1)?

-[Unknown]

jpd002 replied:
My guess is that the answer is yes to both question, even though I haven't tested. My goal was only to test handler ordering and I was thinking such behavior would be better tested somewhere else.

unknownbrackets replied:
Well, I feel like this is an ordering related question - since this is testing ordering/cancelation already, what values actually cause this cancelation?

I could definitely see some game returning an error code from another function here, one that is maybe 0x8xxxxxxx, and the handler always canceling for negative values... but an emulator specifically checking -1, which is likely to be the most common value by far.

-[Unknown]

unknownbrackets commented on tests/kernel/ee/interrupt/intc_order.cpp:
```diff
@@ -0,0 +1,160 @@
+}
+
+void addTestIntcHandler(IntcHandlerFunction handler, s32 next, void *arg) {
+	int handlerId = AddIntcHandler2(INTC_VBLANK_S, handler, next, arg);
```
I notice here you test `AddIntcHandler2` not `AddIntcHandler`, and the other place it's `AddDmacHandler` not `AddDmacHandler2`.  Should we maybe test both in each?

-[Unknown]

jpd002 replied:
`AddDmacHandler2` is not exposed by PS2SDK for some reason. I was using `AddIntcHandler2` for the convenience of being able to specify user data, but I could change the INTC handler test to make it use `AddIntcHandler` (using the same template trick for specifying ids) and make it consistent with the DMAC handler test. 

My goal here is to test handler ordering and not the functions themselves. I think the functions should be tested in a different test.

unknownbrackets replied:
Okay... I think it makes some sense to test the other set of functions elsewhere (although I personally would template it or something and test both here), but I think they should agree within this test, imho.

-[Unknown]

unknownbrackets commented:
Sorry for taking so long to look at this.

-[Unknown]

jpd002 commented:
No problems, thanks for taking time to review this!

jpd002 committed:
Cleanup.

jpd002 committed:
Improved INTC and DMAC handler tests.

- Test return values other than 0 and -1.
- Test both AddIntcHandler and AddIntcHandler2 for INTC.

jpd002 commented:
I've made the changes to style and added some new tests to verify what happens when values other than 0 or -1 are returned by handlers.
